src/layer-model.js

```diff
@@ -1,12 +1,28 @@
 import isEqual from 'lodash/isEqual';
 
+// decodes
+import canvasDecodes from './canvas-decodes';
+
 class LayerModel {
   opacity = 1;
 
   visibility = true;
 
   constructor(layerSpec = {}) {
-    Object.assign(this, layerSpec, { changedAttributes: {} });
+    // parse params before setting model's properties
+    Object.assign(this, {
+      ...layerSpec,
+      ...{
+        ...LayerModel.parse(layerSpec),
+        // compatibility
+        ...layerSpec.params && { params: layerSpec.params },
+        ...layerSpec.sqlParams && { sqlParams: layerSpec.sqlParams },
+        ...layerSpec.decodeParams && { decodeParams: layerSpec.decodeParams },
+        // adds decodeFunction if the layerSpec contains decode_config params (it is a canvas layer)
+        ...Object.prototype.hasOwnProperty.call(layerSpec.layerConfig, 'decode_config') &&
+          { decodeFunction: canvasDecodes[layerSpec.id] }
```

Comment:
Just check that the `canvasDecodes[layerSpec.id]` exists. If not, show a console warning or error

Reply:
Not all layers will need the decoder (basically only canvas layers), if we do that, then the warning will appear always for non-canvas layers.

I would keep it and add below if `decodeFunction` exists and has value, otherwise trigger a warning telling the specified decoder doesn't exist.